:bug:

```
core/src/main/java/jenkins/model/identity/IdentityRootAction.java:[9,34] error: package org.jenkinsci.remoting.util does not exist
core/src/main/java/jenkins/model/identity/IdentityRootAction.java:[9,34] error: package org.jenkinsci.remoting.util does not exist
core/src/main/java/jenkins/model/identity/IdentityRootAction.java:[79,36] error: cannot find symbol
```

--- a/core/src/main/java/jenkins/model/identity/IdentityRootAction.java
+++ b/core/src/main/java/jenkins/model/identity/IdentityRootAction.java
@@ -2,11 +2,12 @@ package jenkins.model.identity;
 
 import hudson.Extension;
 import hudson.model.UnprotectedRootAction;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
 import java.security.interfaces.RSAPrivateKey;
 import java.security.interfaces.RSAPublicKey;
 import org.apache.commons.codec.Charsets;
 import org.apache.commons.codec.binary.Base64;
-import org.jenkinsci.remoting.util.KeyUtils;
 
 /**
  * A simple root action that exposes the public key to users so that they do not need to search for the
@@ -76,6 +77,25 @@ public class IdentityRootAction implements UnprotectedRootAction {
         InstanceIdentityProvider<RSAPublicKey, RSAPrivateKey> provider =
                 InstanceIdentityProvider.get(RSAPrivateKey.class);
         RSAPublicKey key = provider == null ? null : provider.getPublicKey();
-        return key == null ? null : KeyUtils.fingerprint(key);
+        if (key == null) {
+            return null;
+        }
+        // TODO replace with org.jenkinsci.remoting.util.KeyUtils once JENKINS-36871 changes are merged
+        try {
+            MessageDigest digest = MessageDigest.getInstance("MD5");
+            digest.reset();
+            byte[] bytes = digest.digest(key.getEncoded());
+            StringBuilder result = new StringBuilder(Math.max(0, bytes.length * 3 - 1));
+            for (int i = 0; i < bytes.length; i++) {
+                if (i > 0) {
+                    result.append(':');
+                }
+                int b = bytes[i] & 0xFF;
+                result.append(Character.forDigit((b>>4)&0x0f, 16)).append(Character.forDigit(b&0xf, 16));
+            }
+            return result.toString();
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalStateException("JLS mandates MD5 support");
+        }
     }
 }

--- a/core/src/main/java/jenkins/model/identity/InstanceIdentityProvider.java
+++ b/core/src/main/java/jenkins/model/identity/InstanceIdentityProvider.java
@@ -30,6 +30,7 @@ import java.security.PrivateKey;
 import java.security.PublicKey;
 import java.security.cert.X509Certificate;
 import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
 
 /**
  * A source of instance identity.
@@ -43,17 +44,21 @@ public abstract class InstanceIdentityProvider<PUB extends PublicKey, PRIV exten
     /**
      * Gets the {@link KeyPair} that comprises the instance identity.
      *
-     * @return the {@link KeyPair} that comprises the instance identity.
+     * @return the {@link KeyPair} that comprises the instance identity. {@code null} could technically be returned in
+     * the event that a keypair could not be generated, for example if the specific key type of this provider
+     * is not permitted at the required length by the JCA policy.
      */
-    @CheckForNull
+    @Nullable
     public abstract KeyPair getKeyPair();
 
     /**
      * Shortcut to {@link KeyPair#getPublic()}.
      *
-     * @return the public key.
+     * @return the public key. {@code null} could technically be returned in the event that a keypair could not be
+     * generated, for example if the specific key type of this provider is not permitted at the required length by
+     * the JCA policy.
      */
-    @CheckForNull
+    @Nullable
     public PUB getPublicKey() {
         KeyPair keyPair = getKeyPair();
         return keyPair == null ? null : (PUB) keyPair.getPublic();
@@ -62,21 +67,25 @@ public abstract class InstanceIdentityProvider<PUB extends PublicKey, PRIV exten
     /**
      * Shortcut to {@link KeyPair#getPrivate()}.
      *
-     * @return the private key.
+     * @return the private key. {@code null} could technically be returned in the event that a keypair could not be
+     * generated, for example if the specific key type of this provider is not permitted at the required length by
+     * the JCA policy.
      */
-    @CheckForNull
+    @Nullable
     public PRIV getPrivateKey() {
         KeyPair keyPair = getKeyPair();
-        return keyPair == null ? null : (PRIV) keyPair.getPublic();
+        return keyPair == null ? null : (PRIV) keyPair.getPrivate();
     }
 
     /**
      * Gets the self-signed {@link X509Certificate} that is associated with this identity. The certificate
      * will must be currently valid. Repeated calls to this method may result in new certificates being generated.
      *
-     * @return the certificate.
+     * @return the certificate. {@code null} could technically be returned in the event that a keypair could not be
+     * generated, for example if the specific key type of this provider is not permitted at the required length by
+     * the JCA policy.
      */
-    @CheckForNull
+    @Nullable
     public abstract X509Certificate getCertificate();
 
     /**

--- a/core/src/main/java/jenkins/model/identity/InstanceIdentityProvider.java
+++ b/core/src/main/java/jenkins/model/identity/InstanceIdentityProvider.java
@@ -48,17 +48,15 @@ public abstract class InstanceIdentityProvider<PUB extends PublicKey, PRIV exten
      * the event that a keypair could not be generated, for example if the specific key type of this provider
      * is not permitted at the required length by the JCA policy.
      */
-    @Nullable
+    @CheckForNull
     public abstract KeyPair getKeyPair();
 
     /**
      * Shortcut to {@link KeyPair#getPublic()}.
      *
-     * @return the public key. {@code null} could technically be returned in the event that a keypair could not be
-     * generated, for example if the specific key type of this provider is not permitted at the required length by
-     * the JCA policy.
+     * @return the public key. {@code null} if {@link #getKeyPair()} is {@code null}.
      */
-    @Nullable
+    @CheckForNull
     public PUB getPublicKey() {
         KeyPair keyPair = getKeyPair();
         return keyPair == null ? null : (PUB) keyPair.getPublic();
@@ -67,11 +65,9 @@ public abstract class InstanceIdentityProvider<PUB extends PublicKey, PRIV exten
     /**
      * Shortcut to {@link KeyPair#getPrivate()}.
      *
-     * @return the private key. {@code null} could technically be returned in the event that a keypair could not be
-     * generated, for example if the specific key type of this provider is not permitted at the required length by
-     * the JCA policy.
+     * @return the private key. {@code null} if {@link #getKeyPair()} is {@code null}.
      */
-    @Nullable
+    @CheckForNull
     public PRIV getPrivateKey() {
         KeyPair keyPair = getKeyPair();
         return keyPair == null ? null : (PRIV) keyPair.getPrivate();
@@ -81,11 +77,9 @@ public abstract class InstanceIdentityProvider<PUB extends PublicKey, PRIV exten
      * Gets the self-signed {@link X509Certificate} that is associated with this identity. The certificate
      * will must be currently valid. Repeated calls to this method may result in new certificates being generated.
      *
-     * @return the certificate. {@code null} could technically be returned in the event that a keypair could not be
-     * generated, for example if the specific key type of this provider is not permitted at the required length by
-     * the JCA policy.
+     * @return the certificate. {@code null} if {@link #getKeyPair()} is {@code null}.
      */
-    @Nullable
+    @CheckForNull
     public abstract X509Certificate getCertificate();
 
     /**

--- a/core/src/main/resources/hudson/TcpSlaveAgentListener/index.jelly
+++ b/core/src/main/resources/hudson/TcpSlaveAgentListener/index.jelly
@@ -24,6 +24,7 @@ THE SOFTWARE.
 
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
+  <st:contentType value="text/plain;charset=UTF-8"/>
   <!--
     Publicize the TCP port number for JNLP agents so that they know where to conenct.
     Keep the legacy header for better backward compatibility

--- a/test/src/test/java/hudson/TcpSlaveAgentListenerTest.java
+++ b/test/src/test/java/hudson/TcpSlaveAgentListenerTest.java
@@ -1,0 +1,70 @@
+package hudson;
+
+import com.gargoylesoftware.htmlunit.FailingHttpStatusCodeException;
+import com.gargoylesoftware.htmlunit.Page;
+import com.gargoylesoftware.htmlunit.html.HtmlPage;
+import hudson.remoting.Base64;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import javax.annotation.Nullable;
+import jenkins.model.identity.InstanceIdentityProvider;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+public class TcpSlaveAgentListenerTest {
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Test
+    public void headers() throws Exception {
+        r.getInstance().setSlaveAgentPort(-1);
+        try {
+            r.createWebClient().goTo("tcpSlaveAgentListener");
+            fail("Should get 404");
+        } catch (FailingHttpStatusCodeException e) {
+            assertThat(e.getStatusCode(), is(404));
+        }
+        r.getInstance().setSlaveAgentPort(0);
+        Page p = r.createWebClient().goTo("tcpSlaveAgentListener", "text/plain");
+        assertThat(p.getWebResponse().getResponseHeaderValue("X-Instance-Identity"), notNullValue());
+    }
+
+    @TestExtension // TODO remove once instance-identity with provider impl embedded in war
+    public static class ProviderImpl extends InstanceIdentityProvider<RSAPublicKey,RSAPrivateKey> {
+
+        private final KeyPair keyPair;
+
+        public ProviderImpl() throws NoSuchAlgorithmException, InvalidKeySpecException {
+            KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
+            generator.initialize(1024);
+            this.keyPair = generator.generateKeyPair();
+        }
+
+        @Nullable
+        @Override
+        public KeyPair getKeyPair() {
+            return keyPair;
+        }
+
+        @Nullable
+        @Override
+        public X509Certificate getCertificate() {
+            return null;
+        }
+    }
+}

